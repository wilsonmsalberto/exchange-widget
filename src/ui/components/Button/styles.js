import { css } from '@emotion/core';
import styled from '@emotion/styled';

export const StyledButton = styled.button`
  display: block;
  text-align: center;
  text-decoration: none;
  font-size: 1.6rem;
  border-radius: 2.5rem;
  padding: 1rem;
  margin: 0;
  color: #fff;
  border: 0;
  cursor: pointer;

  ${(props) => {
    if (props.green) {
      return css`
        border: none;
        background: #1abc9c;
        color: white;
      `;
    }

    if (props.red) {
      return css`
        border: none;
        background: #e74c3c;
        color: white;
      `;
    }

    if (props.blue) {
      return css`
        border: none;
        background: #3498db;
        color: white;
      `;
    }

    if (props.transparent) {
      return css`
        border: none;
        color: #2c3e50;
      `;
    }

    return css`
      background: transparent;
    `;
  }};

  &:hover {
    transform: translateY(1px);
    box-shadow: 0 2px 3px rgba(0, 0, 0, 0.15);
  }
`;

export default StyledButton;
