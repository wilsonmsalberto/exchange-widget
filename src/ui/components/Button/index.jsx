import React from 'react';
import PropTypes from 'prop-types';

import { StyledButton } from './styles';

const Button = ({
  children,
  onClick,
  type,
  ...otherProps
}) => {
  const handleOnClick = () => {
    if (typeof onClick === 'function') onClick();
  };

  return (
    <StyledButton
      onClick={ handleOnClick }
      type={ type }
      { ...otherProps }
    >
      {children}
    </StyledButton>
  );
};

Button.propTypes = {
  children: PropTypes.node.isRequired,
  onClick: PropTypes.func,
  type: PropTypes.string,
};

Button.defaultProps = {
  onClick: undefined,
  type: 'button',
};

export default Button;
