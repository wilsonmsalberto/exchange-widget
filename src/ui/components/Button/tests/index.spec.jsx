import React from 'react';
import {
  render,
  fireEvent,
} from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import { matchers } from 'jest-emotion';

import Button from '..';

expect.extend(matchers);

const getRenderWrapper = (props) => render(<Button { ...props } />);

describe('Button Component', () => {
  describe('Rendering', () => {
    it('should contain text content', () => {
      const wrapper = getRenderWrapper({ children: 'Button Text' });

      const { asFragment } = wrapper;

      expect(asFragment()).toMatchSnapshot();
    });

    it('should contain node content', () => {
      const MockedNode = <div>Node Content</div>;
      const wrapper = getRenderWrapper({ children: MockedNode });

      const { asFragment } = wrapper;

      expect(asFragment()).toMatchSnapshot();
    });

    it('should contain a button of a different type', () => {
      const wrapper = getRenderWrapper({
        children: 'Button Text',
        type: 'submit',
      });

      const { getByText } = wrapper;

      const button = getByText('Button Text');

      expect(button).toHaveAttribute('type', 'submit');
    });

    it('should render a blue button', () => {
      const wrapper = getRenderWrapper({
        children: 'Blue Button',
        blue: true,
      });

      const { getByText } = wrapper;

      const button = getByText('Blue Button');

      expect(button).toHaveStyleRule('background', '#3498db');
    });

    it('should render a red button', () => {
      const wrapper = getRenderWrapper({
        children: 'Red Button',
        red: true,
      });

      const { getByText } = wrapper;

      const button = getByText('Red Button');

      expect(button).toHaveStyleRule('background', '#e74c3c');
    });

    it('should render a green button', () => {
      const wrapper = getRenderWrapper({
        children: 'Green Button',
        green: true,
      });

      const { getByText } = wrapper;

      const button = getByText('Green Button');

      expect(button).toHaveStyleRule('background', '#1abc9c');
    });

    it('should render a transparent button', () => {
      const wrapper = getRenderWrapper({
        children: 'Transparent Button',
        transparent: true,
      });

      const { getByText } = wrapper;

      const button = getByText('Transparent Button');

      expect(button).toHaveStyleRule('color', '#2c3e50');
    });
  });

  describe('Behavior', () => {
    it('should trigger an onClick function', () => {
      const mockOnClick = jest.fn();

      const wrapper = getRenderWrapper({
        children: 'Button Text',
        onClick: mockOnClick,
      });

      const { getByText } = wrapper;

      const button = getByText('Button Text');

      fireEvent.click(button);

      expect(mockOnClick).toHaveBeenCalledTimes(1);
    });

    it('should do nothing without an onClick function', () => {
      const wrapper = getRenderWrapper({
        children: 'Button Text',
      });

      const { asFragment, getByText } = wrapper;

      const button = getByText('Button Text');

      fireEvent.click(button);

      expect(asFragment()).toMatchSnapshot();
    });
  });
});
