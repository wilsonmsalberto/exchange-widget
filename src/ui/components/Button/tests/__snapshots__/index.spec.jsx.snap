// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Component Behavior should do nothing without an onClick function 1`] = `
<DocumentFragment>
  <button
    class="css-1rdg8jg"
    type="button"
  >
    Button Text
  </button>
</DocumentFragment>
`;

exports[`Button Component Rendering should contain node content 1`] = `
<DocumentFragment>
  <button
    class="css-1rdg8jg"
    type="button"
  >
    <div>
      Node Content
    </div>
  </button>
</DocumentFragment>
`;

exports[`Button Component Rendering should contain text content 1`] = `
<DocumentFragment>
  <button
    class="css-1rdg8jg"
    type="button"
  >
    Button Text
  </button>
</DocumentFragment>
`;
